File: 0875-koko-eating-bananas/0875-koko-eating-bananas.cpp
class Solution {
public:
    int findMax(vector<int> &piles) {
        int ans = 0;
        for(int i=0;i<piles.size();i++) {
            ans = max(ans, piles[i]);
        }
        return ans;
    }

    long long calcTotalHours(vector<int> &piles, int h) {
        long long ans = 0;
        for(int i=0;i<piles.size();i++) {
            ans += ceil((double)piles[i] / (double)h);
        }
        return ans;
    }
    int minEatingSpeed(vector<int>& piles, int h) {
        int low = 1;
        int high = findMax(piles);
        while(low <= high) {
            int mid = low + (high - low) / 2;
            long long totalHours = calcTotalHours(piles, mid);
            if(totalHours <= h) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return low;
    }
};

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.go
func _max(arr []int) int {
    res := arr[0]
    for i := range arr {
        res = max(res, arr[i])
    }
    return res
}

func ok(mid int, p []int) int {
    res := 0
    for i := range p {
        res += int(math.Ceil(float64(p[i])/float64(mid)))
    }
    return res
}

func minEatingSpeed(p []int, h int) int {
    maxi := _max(p)
    l, r := 0, maxi
    ans := maxi

    for l <= r {
        mid := l + (r - l) / 2
        if ok(mid, p) <= h {
            ans = mid
            r = mid - 1
        } else {
            l = mid + 1
        }
    }
    
    return ans
}